### PR TITLE
iOS 2.0

### PR DESCRIPTION
**Pull request by ErisMik**
Check is failing cause cocoapods package isn't out yet. See previous commit for successful appcenter run.

### CODE REVIEW

File: demo/ios/BackgroundService/PorcupineBackgroundServiceDemo/ViewController.swift [hunk 1]
```diff
@@ -28,34 +29,35 @@ class ViewController: UIViewController, UITextViewDelegate {
...
+                self.porcupineManager = try PorcupineManager(accessKey: accessKey, keyword: wakeWord, onDetection: keywordCallback)
+                
+                
                 try porcupineManager.start()
-            } catch {
-                showAlert(message: "Failed to start Porcupine Manager")
+                isRecording = true
+                startButton.setTitle("STOP", for: UIControl.State.normal)
+            } catch let error as PorcupineError {
+                showAlert(message: "Failed to initialize and start Porcupine:\n\(error)")
```
Comment (laves) [inline demo/ios/BackgroundService/PorcupineBackgroundServiceDemo/ViewController.swift:50]: We should be more descriptive with this init error and disable the start button once we determine they haven't put in a valid AccessKey. See the Octopus iOS demo for how I catch the individual exceptions and show them to the user.

File: demo/ios/BackgroundService/PorcupineBackgroundServiceDemo/ViewController.swift [hunk 2]
```diff
@@ -28,34 +29,35 @@ class ViewController: UIViewController, UITextViewDelegate {
...
+                showAlert(message: "Something went wrong: \(error)")
                 return
             }
 
-            isRecording = true
-            startButton.setTitle("STOP", for: UIControl.State.normal)
+
         } else {
-            porcupineManager.stop()
+            if porcupineManager != nil {
```
Comment (laves) [inline demo/ios/BackgroundService/PorcupineBackgroundServiceDemo/ViewController.swift:58]: At the top we assert that PorcupineManager will have a value (with the `!`). If we ever think it won't have a value, we need to change the type to `?`

File: demo/ios/ForegroundApp/PorcupineForegroundAppDemo/ViewController.swift [hunk 3]
```diff
@@ -13,7 +13,9 @@ import Porcupine
 class ViewController: UIViewController, UIPickerViewDelegate, UIPickerViewDataSource {
     @IBOutlet weak var wakeWordPicker: UIPickerView!
     @IBOutlet weak var startButton: UIButton!
+    @IBOutlet weak var errorLabel: UILabel!
 
+    let accessKey = "" // Obtained from Picovoice Console (https://console.picovoice.ai)
```
Comment (laves) [inline demo/ios/ForegroundApp/PorcupineForegroundAppDemo/ViewController.swift:18]: ${YOUR_ACCESS_KEY_HERE}

File: demo/ios/ForegroundApp/PorcupineForegroundAppDemo/ViewController.swift [hunk 4]
```diff
@@ -13,7 +13,9 @@ import Porcupine
 class ViewController: UIViewController, UIPickerViewDelegate, UIPickerViewDataSource {
     @IBOutlet weak var wakeWordPicker: UIPickerView!
     @IBOutlet weak var startButton: UIButton!
+    @IBOutlet weak var errorLabel: UILabel!
```
Comment (laves) [inline demo/ios/ForegroundApp/PorcupineForegroundAppDemo/ViewController.swift:16]: label should be above the start button to keep it consistent with other demos

File: demo/ios/ForegroundApp/PorcupineForegroundAppDemo/ViewController.swift [hunk 5]
```diff
@@ -69,25 +75,32 @@ class ViewController: UIViewController, UIPickerViewDelegate, UIPickerViewDataSo
...
-                porcupineManager = try PorcupineManager(keyword: wakeWord, onDetection: keywordCallback)
+                porcupineManager = try PorcupineManager(accessKey: accessKey, keyword: wakeWord, onDetection: keywordCallback, errorCallback: errorCallback)
                 try porcupineManager.start()
+                
+                errorLabel.isHidden = true
+                wakeWordPicker.isUserInteractionEnabled = false
+                isRecording = true
+                startButton.setTitle("STOP", for: UIControl.State.normal)
+            } catch let error as PorcupineError {
+                errorLabel.isHidden = false
```
Comment (laves) [inline demo/ios/ForegroundApp/PorcupineForegroundAppDemo/ViewController.swift:93]: disable controls (visibility as well) when you have an init error

File: demo/ios/ForegroundApp/PorcupineForegroundAppDemo/ViewController.swift [hunk 6]
```diff
@@ -69,25 +75,32 @@ class ViewController: UIViewController, UIPickerViewDelegate, UIPickerViewDataSo
...
+                errorLabel.text = "Something went wrong"
                 return
             }
-
-            wakeWordPicker.isUserInteractionEnabled = false
-            isRecording = true
-            startButton.setTitle("STOP", for: UIControl.State.normal)
         } else {
-            porcupineManager.stop()
+            if porcupineManager != nil {
```
Comment (laves) [inline demo/ios/ForegroundApp/PorcupineForegroundAppDemo/ViewController.swift:101]: can it be null?

File: demo/ios/BackgroundService/PorcupineBackgroundServiceDemo/ViewController.swift [hunk 7]
```diff
@@ -28,34 +29,35 @@ class ViewController: UIViewController, UITextViewDelegate {
...
-        do {
-            self.porcupineManager = try PorcupineManager(keyword: wakeWord, onDetection: keywordCallback)
-        } catch {
-            showAlert(message: "Failed to initialize Porcupine Manager")
-        }
     }
 
     @IBAction func toggleStartButton(_ sender: UIButton) {
         if !isRecording {
-            NotificationManager.shared.requestNotificationAuthorization()
```
Comment (laves) [inline demo/ios/BackgroundService/PorcupineBackgroundServiceDemo/ViewController.swift:46]: Did we not need this?

File: binding/ios/README.md [hunk 8]
```diff
@@ -49,8 +58,11 @@ To create an instance of PorcupineManager pass the desired keyword to the constr
 ```swift
 import Porcupine
 
+let accessKey = "" // Obtained from Picovoice Console (https://console.picovoice.ai)
+
 do {
     PorcupineManager porcupineManager = try PorcupineManager(
```
Comment (laves) [inline binding/ios/README.md:64]: my typo, but this should be `let porcupineManager = try PorcupineManager(`

**Review by laves — CHANGES_REQUESTED** [review]

File: demo/ios/BackgroundService/PorcupineBackgroundServiceDemo/ViewController.swift (hunk 2 above)
Comment (ErisMik) [reply demo/ios/BackgroundService/PorcupineBackgroundServiceDemo/ViewController.swift:58]: I was noticing when the `PorcupineManager` constructor failed it was left without a value, but I guess with the changes to how we are handling error's it's okay.

**Review by laves — APPROVED** [review]

**Pull request merged**